File: EmailServices/src/utils/retry.util.ts
// src/utils/retry.util.ts

/**
 * Calculates an exponential backoff delay based on retry count.
 * Example: 0 → 2s, 1 → 4s, 2 → 8s, up to 32s max.
 */
export function calculateBackoff(retryCount: number): number {
  const baseDelay = 2000; // 2 seconds
  const maxDelay = 32000; // 32 seconds
  const delay = baseDelay * 2 ** retryCount;
  return Math.min(delay, maxDelay);
}

File: PushNotificationService/src/main/java/dev/slethware/pushnotifications/util/RetryUtil.java
package dev.slethware.pushnotifications.util;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

@Component
public class RetryUtil {

    private final long baseDelay;
    private final long maxDelay;

    public RetryUtil(
            @Value("${retry.base-delay-ms}") long baseDelay,
            @Value("${retry.max-delay-ms}") long maxDelay) {
        this.baseDelay = baseDelay;
        this.maxDelay = maxDelay;
    }

    // Calculates exponential backoff delay.
    public long calculateBackoff(int retryCount) {
        long delay = baseDelay * (long) Math.pow(2, retryCount);
        return Math.min(delay, maxDelay);
    }
}